vdom/core.py

```diff
@@ -56,157 +53,95 @@ def to_json(el, schema=None):
 
     return json_el
 
-
 class VDOM(object):
     """A basic virtual DOM class which allows you to write literal VDOM spec
 
-    >>> VDOM({ 'tagName': 'h1', 'children': 'Hey', 'attributes': {}})
-
-    It's probably better to use `vdom.h` or the helper functions:
-
-    >>> from vdom import h
-    >>> h('h1', 'Hey')
-    <h1 />
+    >>> VDOM(tag_name='h1', children='Hey', attributes: {}})
 
     >>> from vdom.helpers import h1
     >>> h1('Hey')
-
     """
-    _schema = VDOM_SCHEMA
-    _obj = None
+    # This class should only have these 4 attributes
+    __slots__ = ['tag_name', 'attributes', 'children', 'key']
 
-    def __init__(self, obj, schema=None):
-        # we need to assign self.schema first,
-        # because it is used to validate the object
+    def __init__(self, tag_name, attributes=None, children=None, key=None, schema=None):
         if schema is not None:
-            self.schema = schema
-        self.obj = obj
+            warnings.warn('Passing schema param to VDOM constructor is deprecated')
```

Comment:
I'll bet this is where @mpacer would take exception. Why would you want to deprecate the schema param?

Now I remember my conversation with M about the schema and some major features and drawbacks:

Pro: People can pass in a schema that enforces particular attributes are set (adhering to current w3c standards, accessibility)

Con: Inevitably some person or library will write a schema that is no longer valid with the core schema and for which we get bug reports "Hey what's wrong with vdom" when in fact it's coming from somewhere else.

Reply:
I actually looked to see if the current svg / helper modules were using schema, and noticed they weren't...

Since the `VDOM(<somedict>)` form is deprecated here, JSON Schema is no longer a useful way of enforcing schemas. We're sortof enforcing that schema with Python instead, and the `allow_children` and `tag_name` attribute in create_component. This is a super small subset of what JSON Schema can do, ofc. But it's guaranteed to produce things that are valid vdom.

If someone wants to do such validation, there are a few options (in increasing order of intrusiveness):

1. They do this validation on emit, by validating output of `to_dict`
2. We add a convenience `validate` method that users *must* pass a schema to for validation (since we guarantee that we already are valid vdom)
3. validate ourselves recursively in the constructor.

IMO, this is a 'YAGNI' problem that we can fix when we actually run into it. Depending on *why* someone wants schemas, any of the 1-3 solutions are the right thing to do. The current design doesn't invalidate any of them.

However, since I removed this 'schema' functionality here, am happy to re-add here to be fully backwards compatible :)

Reply:
From @mpacer's comments in the PR, looks like I'll need to add the custom schema support back. I shall do so.

Reply:
@rgbkrk Well predicted. As I said in a more general comment (and I still need to work through this a bit) but one of my aims is to enable faster ways to compose what are effectively presentational components out of VDOM elements and that will be a lot easier if we can pass in and validate a custom schema against VDOM, in particular by putting constraints on attributes, children, etc.

As far as the concern about the bug reports… we could actually also check that the schema passes whatever the core schema is at the time in addition to whatever custom schema is being passed in. That way it always a subset of the full VDOM schema.

Also, this is making me realise that we should have a way to pass in different versions of the schema here as the base schema (but please let's make it optional unlike `nbformat`'s read and write `as_version` stuff)?